raster/src/main/scala/geotrellis/raster/render/BreakMap.scala

```diff
@@ -0,0 +1,133 @@
+  */
+class BreakMap[
+  @specialized(Int, Double) A: Order,
+  @specialized(Int, Double) B: Order
+](
+  breakMap: Map[A, B],
+  strategy: MapStrategy[B],
+  noDataCheck: A => Boolean
+) extends (A => B) with Serializable {
+
+  /* A Binary Tree of the mappable values */
+  private lazy val vmTree: BTree[(A, B)] = {
+    val a: Array[(A, B)] = breakMap.toArray
+
+    Sorting.quickSort(a)
+
+    BTree.fromSortedSeq(a.toIndexedSeq).get
+  }
+
+  /* Yield a btree search predicate function based on boundary type options. */
+  private val branchPred: (A, BTree[(A, B)]) => Either[Option[BTree[(A, B)]], (A, B)] = {
+    strategy.boundary match {
+      case LessThan => { (z, tree) => tree match {
+        case BTree(v, None, _)    if z < v._1                       => Right(v)
+        case BTree(v, Some(l), _) if z < v._1 && z >= l.greatest._1 => Right(v)
+        case BTree(v, l, _)       if z < v._1                       => Left(l)
+        case BTree(_, _, r)                                         => Left(r)
+      }}
+      case LessThanOrEqualTo => { (z, tree) => tree match {
+        case BTree(v, None, _)    if z <= v._1                      => Right(v)
+        case BTree(v, Some(l), _) if z <= v._1 && z > l.greatest._1 => Right(v)
+        case BTree(v, l, _)       if z < v._1                       => Left(l)
+        case BTree(_, _, r)                                         => Left(r)
+      }}
+      case Exact => { (z, tree) => tree match { /* Vanilla Binary Search */
+        case BTree(v, _, _) if z == v._1 => Right(v)
+        case BTree(v, l, _) if z < v._1  => Left(l)
+        case BTree(_, _, r)              => Left(r)
+      }}
+      case GreaterThanOrEqualTo => { (z, tree) => tree match {
+        case BTree(v, _, None)    if z >= v._1                    => Right(v)
+        case BTree(v, _, Some(r)) if z >= v._1 && z < r.lowest._1 => Right(v)
+        case BTree(v, l, _)       if z < v._1                     => Left(l)
+        case BTree(_, _, r)                                       => Left(r)
+      }}
+      case GreaterThan => { (z, tree) => tree match {
+        case BTree(v, _, None)    if z > v._1                     => Right(v)
+        case BTree(v, _, Some(r)) if z > v._1 && z <= r.lowest._1 => Right(v)
+        case BTree(v, l, _)       if z <= v._1                    => Left(l) /* (<=) is correct here! */
+        case BTree(_, _, r)                                       => Left(r)
+      }}
+    }
+  }
+
+  def apply(z: A): B = {
+    if (noDataCheck(z)) {
+      strategy.noDataValue
+    } else {
+      vmTree.searchWith(z, branchPred) match {
+        case Some((_, v)) => v
+        case None if strategy.strict => sys.error(s"Value ${z} did not have an associated value.")
+        case _ => strategy.fallbackValue
+      }
+    }
+  }
+}
+
+/** Helper methods for constructing BreakMaps. */
+object BreakMap {
+  def i2i(m: Map[Int, Int], s: MapStrategy[Int] = MapStrategy.int): BreakMap[Int, Int] =
```

Comment:
Although I'm opposed to overloads in general, after thinking about it, overloaded `apply` isn't that terrible so long as the return type matches the parent class, which it would here.

I'll switch to `apply`.